src/fock.jl

```diff
@@ -21,44 +21,63 @@ end
 ==(b1::FockBasis, b2::FockBasis) = (b1.N==b2.N && b1.offset==b2.offset)
 
 """
-    number(b::FockBasis)
+    number([T=ComplexF64,] b::FockBasis)
 
-Number operator for the specified Fock space.
+Number operator for the specified Fock space with optional data type `T`.
 """
-function number(b::FockBasis)
-    diag = complex.(b.offset:b.N)
+function number(::Type{T}, b::FockBasis) where T
+    diag = T.(b.offset:b.N)
     data = spdiagm(0 => diag)
     SparseOperator(b, data)
 end
 
+number(b::FockBasis) = number(ComplexF64, b)
+
+
 """
-    destroy(b::FockBasis)
+    destroy([T=ComplexF64,] b::FockBasis)
 
-Annihilation operator for the specified Fock space.
+Annihilation operator for the specified Fock space with optional data type `T`.
 """
-function destroy(b::FockBasis)
-    diag = complex.(sqrt.(b.offset+1.:b.N))
+function destroy(::Type{C}, b::FockBasis) where C
+    T = real(C)
```

Comment:
Could you explain why you're taking the detour over `real(C)` here?

Reply:
Sorry I was a bit busy today. I always struggle with this issue. One major concern I usually have is to get the right type in computations, especially when using extended precision datatypes (the reason why I have proposed these changes in the first place). 

Assuming that one specifies the datatype as `Complex{T}` for some `T` we still needs to compute the `sqrt`s with the correct precision, as by default `sqrt` uses `Float64` if the input type is not an `AbstractFloat`.

Another path we could follow, is to always assume the specified type is real and construct the state/operator with the correct complex type.

Reply:
Thanks for the explanation, that makes sense.